src/kolibri_gnome/desktop_launcher/application.py

```diff
@@ -491,3 +695,100 @@ def open_kolibri_home(self):
     def quit(self):
         for window in self.__windows:
             window.close()
+
+
+class GenericApplication(Application):
+    def _create_window(self, target_url=None):
+        return KolibriGenericWindow(target_url, delegate=self)
+
+
+class ChannelApplication(Application):
+    def __init__(self, channel_id, *args, **kwargs):
+        self.__channel_id = channel_id
+        super().__init__(*args, **kwargs)
+
+    @property
+    def channel_id(self):
+        return self.__channel_id
+
+    @property
+    def default_url(self):
+        return "x-kolibri-app:/learn#topics/{channel_id}".format(
+            channel_id=self.channel_id
+        )
+
+    def _create_window(self, target_url=None):
+        return KolibriChannelWindow(self.channel_id, target_url, delegate=self)
+
+    def url_is_valid(self, url):
+        return super().url_is_valid(url) and self.__is_url_in_channel(url)
+
+    def open_url_in_external(self, url):
+        if self.is_kolibri_url(url):
+            url = self.url_to_x_kolibri_app(url)
+        self.open_in_browser(url)
+
+    def __is_url_in_channel(self, url):
+        url_tuple = urlsplit(url)
+
+        if re.match(r"^\/(?P<lang>\w+\/)?learn\/?", url_tuple.path):
+            return self.__is_learn_fragment_in_channel(url_tuple.fragment)
+        elif re.match(r"^\/(?P<lang>\w+\/)?user\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/(?P<lang>\w+\/)?logout\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/(?P<lang>\w+\/)?redirectuser\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/zipcontent\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/app\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/static\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/downloadcontent\/?", url_tuple.path):
+            return True
+        elif re.match(r"^\/content\/storage\/?", url_tuple.path):
```

Comment:
This can be simplified grouping regular expressions:
```suggestion
        elif re.match(r"^\/(?P<lang>\w+\/)?(user|logout|redirectuser)\/?", url_tuple.path):
            return True
        elif re.match(r"^\/(zipcontent|app|static|downloadcontent|content\/storage)\/?", url_tuple.path):
            return True
```

But maybe this way is clearer.

Reply:
Ah, I like it. I'd started off copying these from Kolibri's urls.py, but seeing as this code is nowhere near that code, the similarity between the two doesn't really help anything. That's much better! Thank you :)